Fix EZP-28099: Wrong Urls at equal beginning path prefixes

blankse opened the pull request:
JIRA: https://jira.ez.no/browse/EZP-28099

If you have 2 path prefixes that starts with the same name.
e.g.
- eZ Platform
- eZ Platform Enterprise

And generate a link from the short path prefix to the large. The url is generate without leading slash and with half path prefix.
e.g.
`-Enterprise/Imprint`
instead of
`/eZ-Platform-Enterprise/Imprint`

andrerom commented:
would need some additional tests for this, @alongosz maybe you have an idea where that might fit?

alongosz commented:
Hello, sorry for the delay, busy Sprint ;)
@blankse I'm having trouble reproducing this issue.
Do I understand correctly that you have under root folder, two following subfolders:
- eZ Platform
- eZ Platform Enterprise

and under the second one you have e.g. Article called Imprint?

If so, what is the next step that results in this wrong URL?

blankse commented:
You need a siteaccess with root location "eZ Platform".
Then show there a link to an article under "eZ Platform Enterprise".
There is outside from the own root location. But with excluded_uri_prefixes this should be possible.

alongosz commented:
> You need a siteaccess with root location "eZ Platform".
> Then show there a link to an article under "eZ Platform Enterprise".
> There is outside from the own root location. But with excluded_uri_prefixes this should be possible.

Ok, I was able to confirm that. I'll take a closer look at the solution and proper place to improve tests as soon as possible.

blankse commented:
I rebased and created a issue: https://jira.ez.no/browse/EZP-28099

blankse commented:
Ping @andrerom @alongosz 
Should I rebase to 6.7?

alongosz commented:
@blankse Yes you could, but we're still missing some test case. Sorry I wasn't able to help you with this, but my work queue has been full for several months now.

andrerom commented:
@blankse yes, 6.7 is good for this one.

As for tests, you can add a new test case to cover this in:
https://github.com/ezsystems/ezpublish-kernel/blob/9fb850707a66901b0d627e132fe8395c5844d2b9/eZ/Publish/Core/MVC/Symfony/Routing/Tests/UrlAliasGeneratorTest.php

As usual make sure it fails before this fix so it captures the issue ;)

blankse commented:
@andrerom @alongosz I rebased to 6.7 and added an test case .

It failed without the patch:
```
There was 1 failure:

1) eZ\Publish\Core\MVC\Symfony\Routing\Tests\UrlAliasGeneratorTest::testDoGenerateRootLocation with data set #9 (eZ\Publish\API\Repository\Values\Content\URLAlias Object (...), false, '/products/ez-publish', '/prod')
Failed asserting that two strings are identical.
--- Expected
+++ Actual
@@ @@
-/products/ez-publish
+ucts/ez-publish

/var/www/ezpublish-kernel/eZ/Publish/Core/MVC/Symfony/Routing/Tests/UrlAliasGeneratorTest.php:358
```

blankse commented:
How long will the QA take? Will it still be part of v7.4.0?

micszo commented:
Yes, I think it will be.

micszo commented:
Verified with sanities on v1.7.8 and v1.13.4.
Also on v2.3.2. And 2.4.0-beta1.

lserwatka commented:
@andrerom can you take over on merging this up to 6.13, 7.3 and master?
